Adding a dynamic questions Bubble tea UI feature into the cobra pkg

sabouaram opened the pull request:
- Adding a dynamic questions Bubble tea UI feature into the cobra pkg
- Description:
This pull request adds a dynamic question Bubble tea UI feature into the `cobra` package, enhancing user interaction. The implementation supports both Unix/Linux and Windows environments, ensuring broad compatibility across platforms.

sabouaram commented:
`package main

import (
	"errors"
	"fmt"
	"os"
	"strconv"

	tea "github.com/charmbracelet/bubbletea"
	"github.com/spf13/cobra"
)

type Question struct {
	Text    string
	Options []string
	Handler func(string) error
}

var (
	Name  string
	Age   int
	Color string
)

func main() {
	var rootCmd = &cobra.Command{
		Use:   "mycli",
		Short: "A CLI application using Cobra and Bubble Tea",
		Run: func(cmd *cobra.Command, args []string) {
			runInteractiveUI()
			fmt.Println(Name, Age, Color)
			fmt.Println("ok")
		},
	}

	if err := rootCmd.Execute(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}

func runInteractiveUI() {
	var questions = []Question{
		{
			Text:    "Select your favorite color:",
			Options: []string{"Red", "Blue", "Green"},
			Handler: handleColorInput,
		},
		{
			Text:    "What is your name: ",
			Handler: handleNameInput,
		},
		{
			Text:    "Enter your age: ",
			Handler: handleAgeInput,
		},
		// Add more questions as needed
	}

	p := tea.NewProgram(model(questions))
	if err := p.Start(); err != nil {
		fmt.Println("Error:", err)
		os.Exit(1)
	}
}

func model(questions []Question) tea.Model {
	return &promptModel{questions: questions, cursor: 0}
}

type promptModel struct {
	questions []Question
	index     int
	input     string
	cursor    int
	errorMsg  string
}

func (m *promptModel) Init() tea.Cmd {
	return nil
}
func (m *promptModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
	var err error
	if m.index >= len(m.questions) {
		return m, tea.Quit
	}
	switch msg := msg.(type) {
	case tea.KeyMsg:
		if msg.String() == "ctrl+c" {
			return m, tea.Quit
		}
		switch msg.Type {
		case tea.KeyEnter:
			if m.index < len(m.questions) {
				if len(m.questions[m.index].Options) > 0 {
					err = m.questions[m.index].Handler(m.questions[m.index].Options[m.cursor])
				} else {
					err = m.questions[m.index].Handler(m.input)
				}
				if err != nil {
					m.errorMsg = err.Error()
					m.input = ""
					m.cursor = 0
					return m, nil
				}
				m.input = ""
				m.index++
				m.cursor = 0
				return m, nil
			} else {
				return nil, tea.Quit
			}
		case tea.KeyDown:
			if len(m.questions[m.index].Options) > 0 {
				m.cursor = (m.cursor + 1) % len(m.questions[m.index].Options)
			}
		case tea.KeyUp:
			if len(m.questions[m.index].Options) > 0 {
				m.cursor = (m.cursor - 1 + len(m.questions[m.index].Options)) % len(m.questions[m.index].Options)
			}
		case tea.KeyTab:

		case tea.KeySpace:

		case tea.KeyBackspace:
			if len(m.input) > 0 {
				m.input = m.input[:len(m.input)-1]
			}
		default:
			m.input += msg.String()
		}
	}

	return m, nil
}

func (m *promptModel) View() string {
	if m.index >= len(m.questions) {
		return ""
	}
	if m.errorMsg != "" {
		return "" + m.errorMsg + "\n" + m.questions[m.index].Text + m.input
	}

	view := m.questions[m.index].Text

	if len(m.questions[m.index].Options) > 0 {
		view += "\n"
		for i, option := range m.questions[m.index].Options {
			cursor := " "
			if i == m.cursor {
				cursor = "→"
			}
			view += fmt.Sprintf("%s %d. %s\n", cursor, i+1, option)
		}
	} else {
		if len(m.input) > 0 {
			view += "" + m.input + "\n"
		} else {
			view += "\n"
		}
	}

	return view
}

func handleNameInput(input string) error {
	Name = input
	return nil
}

func handleColorInput(input string) error {
	Color = input
	return nil
}

func handleAgeInput(input string) error {
	var err error
	Age, err = strconv.Atoi(input)
	if err != nil {
		return errors.New("Please enter a valid age")
	} else {
		return nil
	}
}
`

sabouaram commented:
Features:
- Standalone Usage: Define interactive questionnaires independently of Cobra CLI.
- Integration with Cobra CLI: Seamlessly integrate with Cobra commands for interactive command-line interfaces.
- Dynamic Question Handling: Questions adjust dynamically based on user input and context.
- Pagination Support: Automatically paginate options for improved usability.
- Flexible Control Flow: Choose when to execute the UI based on application control flow.
- Protection for Sensitive Input: Support for password-type questions to securely handle sensitive information.
- Clear Usage Guidelines: Well-documented usage examples provided for various question types.

Usage Examples:
- Choice Questions: Present users with a list of options to choose from.
- Protected Input Questions: Securely handle sensitive information like passwords.
- File Selection Input Questions: Enable users to select files interactively.
- Normal Input Questions: Collect standard user input for various purposes.

Integration with Cobra CLI:
- Easily set up interactive command-line interfaces within Cobra CLI commands.
- Enhance user experience by providing dynamic and interactive prompts during command execution.

nabbar commented on cobra/ui/model.go:
```diff
@@ -0,0 +1,365 @@
+		switch msg.Type {
+		case tea.KeyEnter:
+			if u.index < len(u.questions) {
+				if u.questions[u.index].FilePath {
```
make a sub private function to manage filepath statement like 
`
func (u *ui) questionFilePath() {
...
}
` 
and replace this inline statement to call of this sub function like this: 
`if u.questions[u.index].FilePath {
    u.questionFilePath()
}`

sabouaram replied:
Done

nabbar commented on cobra/ui/model.go:
```diff
@@ -0,0 +1,365 @@
+				return nil, tea.Quit
+			}
+		case tea.KeyDown:
+			if u.questions[u.index].FilePath {
```
move this statement to a private dedicated function

nabbar commented on cobra/ui/model.go:
```diff
@@ -0,0 +1,365 @@
+				u.cursor--
+			}
+		case tea.KeyRight:
+			if u.questions[u.index].FilePath {
```
move this statement to a dedicated statement

nabbar commented on cobra/ui/model.go:
```diff
@@ -0,0 +1,365 @@
+	}
+	switch msg := msg.(type) {
+	case tea.KeyMsg:
+		if msg.String() == "ctrl+c" {
```
use constant instead of fixed string

sabouaram replied:
Done instead I used the tea CTRLC constant

nabbar commented on cobra/ui/model.go:
```diff
@@ -0,0 +1,365 @@
+	}
+	view := u.questions[u.index].Text
+
+	if u.questions[u.index].FilePath {
```
move this statement to dedicated private function

nabbar commented on cobra/ui/model.go:
```diff
@@ -0,0 +1,365 @@
+			view += "Error: " + u.errorMsg + "\n"
+		}
+		fileInfo, err := os.Stat(u.input)
+		if err == nil && fileInfo.IsDir() {
```
if you store in memory a complete list of file, print it completly
if you manage page, read only awaiting page into memory to prevent memory consumption

nabbar commented on cobra/ui/interface.go:
```diff
@@ -0,0 +1,57 @@
+	Options      []string
+	Handler      func(string) error
+	PasswordType bool
+	FilePath     bool
```
add option to custom contents following input
like spefic function by key pressed : 
FuncKeyDown
FuncKeyUp
FuncUpdate...

as that if function are not set, use standard run, but if user defined set, call the user defined function

sabouaram replied:
For instance just the cursor and the question color are allowed to be specified by the user for each question, I think we most keep the the UP - DOWN - BACKSPACE as they are as well or if you agree we can at least let the user to choose them explicitly but the code logic (the behind functionalities will remain intact)

nabbar reviewed (commented):
this MR is great, but could u make some update / evol 
to kindly reading code & allow more capabilities with your package ?

sabouaram commented:
In our recent commits, we've addressed the requested modifications.
Key changes include:
- Refactoring both the view and update functions to enhance user-friendliness. 
- Fine-tuned the logic for listing files. 
- Introduced two new fields in the questions struct to enable users to dynamically select their preferred cursor and color (using fatih/color) for each question.
